### PR TITLE
experimental EV routing request parameters and data

### PR DESCRIPTION
**Pull request by LukasPaczos**
Introduces a set of experimental request and response parameters supporting EV routing capabilities. Not all parameters are yet supported, even on staging endpoints.

All the experimental parameters are subject to change and evolve, therefore, I opted to avoid documenting anything specific for now.

### CODE REVIEW

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/ExperimentalWaypointMetadata.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+    /**
+     * Object representing experimental value.
+     * <p>
+     * All available experimental values are subject to change at any time.
+     *
+      @param powerKw powerKw
+     */
+    @NonNull
+    public abstract Builder powerKw(@Nullable Integer powerKw);
```
Comment (purew) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/ExperimentalWaypointMetadata.java:181]: Question regarding capitalization. Watt is usually symbolized with `W` (capital letter) and kilo by `k` which doesn't translate well to snake case :) Is `powerKiloWatt` too verbose?

I'm likely overanalyzing this. Feel free to ignore me :)

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java [hunk 2]
```diff
@@ -482,6 +487,22 @@ private RouteOptions routeOptions() {
...
       .enableRefresh(true)
       .metadata(true)
+      .experimental(
+        RouteOptions.Experimental.builder()
+          .engine("electric")
+          .evInitialCharge(80000)
+          .evMaxCharge(80000)
+          .evConnectorTypes("ccs_combo_type1,ccs_combo_type2")
+          .energyConsumptionCurve("0,300;20,160;80,140;120,180")
+          .auxiliaryConsumption("auxiliaryConsumption")
```
Comment (purew) [inline services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:497]: The auxilliare consumption is currently not prioritized as that information can be passed through overloaded via the charging-curve parameter https://github.com/mapbox/nav-api-intake/issues/893#issuecomment-969241997

File: services-directions/src/main/java/com/mapbox/api/directions/v5/DirectionsService.java [hunk 3]
```diff
@@ -153,7 +167,7 @@
     @Field("radiuses") String radiuses,
     @Field("steps") Boolean steps,
     @Field("bearings") String bearings,
-    @Query("avoid_maneuver_radius") Integer avoidManeuverRadius,
+    @Field("avoid_maneuver_radius") Integer avoidManeuverRadius,
```
Comment (purew) [inline services-directions/src/main/java/com/mapbox/api/directions/v5/DirectionsService.java:170]: Unrelated change?

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/ExperimentalWaypointMetadata.java (hunk 1 above)
Comment (LukasPaczos) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/ExperimentalWaypointMetadata.java:181]: Yea, this is an unfortunate setup. On top of snake case, code style verification in this project forbids any acronyms in the API signature (more than 2 capital letters next to each other), so `powerKW` cannot be used and I don't think we need to necessarily revisit this now.

I like the `powerKiloWatt` suggestion :+1:

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java [hunk 4]
```diff
@@ -18,7 +18,8 @@
...
+    "https://api.mapbox.com/directions/v5/mapbox/driving/-122.4003312,37.7736941;-122.4187529,37.7689715;-122.4255172,37.7775835?access_token=pk.token&geometries=polyline6&alternatives=false&overview=full&radiuses=;unlimited;5.1&steps=true&avoid_maneuver_radius=200&bearings=0,90;90,0;&layers=-42;;0&continue_straight=false&annotations=congestion,distance,duration&language=ru&roundabout_exits=false&voice_instructions=true&banner_instructions=true&voice_units=metric&exclude=toll,ferry&include=hot,hov2&approaches=;curb;&waypoints=0;1;2&waypoint_names=;two;&waypoint_targets=;12.2,21.2;&enable_refresh=true&walking_speed=5.11&walkway_bias=-0.2&alley_bias=0.75&snapping_include_closures=;false;true&arrive_by=2021-01-01'T'01:01&depart_at=2021-02-02'T'02:02&max_height=1.5&max_width=1.4&metadata=true&engine=electric&ev_initial_charge=80000&ev_max_charge=80000&ev_connector_types=ccs_combo_type1,ccs_combo_type2&energy_consumption_curve=0,300;20,160;80,140;120,180&auxiliary_consumption=auxiliaryConsumption&ev_ascent=5.0&ev_descent=6.0&ev_charging_curve=0,100000;40000,70000;60000,30000;80000,10000&ev_max_ac_charging_power=1&ev_min_charge_at_destination=2&ev_min_charge_at_charging_station=3";
```
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:22]: :+1: unrelated for this PR though since these are unit tests and we mock the result anyway.

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java (hunk 2 above)
Comment (LukasPaczos) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:497]: Would you recommend removing this param for now?

File: services-directions/src/main/java/com/mapbox/api/directions/v5/DirectionsService.java (hunk 3 above)
Comment (LukasPaczos) [reply services-directions/src/main/java/com/mapbox/api/directions/v5/DirectionsService.java:170]: Yes, I can extract it out to a separate commit.

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java [hunk 5]
```diff
@@ -416,4 +416,47 @@ private DirectionsCriteria() {
   })
   public @interface ApproachesCriteria {
   }
+
+  // EXPERIMENTAL
```
Comment (VysotskiVadim) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java:420]: Have we always highlighted experimentation API by a comment? Maybe it's better to add an 'Unstable' or 'Experimental' annotation like [this](https://stackoverflow.com/a/32616239/3599375) ?

Comment (LukasPaczos) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java:420]: This particular comment is just for us to organize the code, unless you mean all the other actual javadoc comments?

There's no precedent for experimental mapbox-java APIs and I quickly tested out adding a custom annotation that uses `@Documented` but it didn't do the trick, javadoc still requires me to write documentation for the annotation method.

Comment (LukasPaczos) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java:420]: Or are you thinking about a supplement in form of annotation, next to javadoc comments saying that the method is experimental?

Comment (VysotskiVadim) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java:420]: I was thinking about an `Experimental` annotation which will be visible in java-docs just as annotation. It could be a standard way to mark an API as not stable instead of manually writing messages about stability

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 6]
```diff
@@ -1871,4 +1979,313 @@ public Builder snappingIncludeClosuresList(@Nullable List<Boolean> snappingClosu
...
+
+  /**
+   * Object representing experimental value.
+   * <p>
+   * All available experimental values are subject to change at any time.
+   */
+  @AutoValue
+  public abstract static class Experimental extends DirectionsJsonObject {
+
+    static String[] experimentalParameters = new String[] {
```
Comment (VysotskiVadim) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:1991]: It seems like the `experimentalParameters` array is used only for lookup if a filed is experimental. Maybe it's easier to keep all the experimental fields name in a `HashSet`?

Comment (LukasPaczos) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:1991]: @VysotskiVadim what would be a practical difference?

Comment (VysotskiVadim) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:1991]: You won't need to implement custom lookup with HashSet. Instead of 
```kotlin
        String property = URLDecoder.decode(query.substring(0, idx), UTF_8);
        String value = URLDecoder.decode(query.substring(idx + 1), UTF_8);

        boolean isExperimental = false;
        for (int i = 0; i < Experimental.experimentalParameters.length; i++) {
          if (property.equals(Experimental.experimentalParameters[i])) {
            isExperimental = true;
            if (!optionsJson.has("experimental")) {
              optionsJson.add("experimental", new JsonObject());
            }
            JsonObject experimental = optionsJson.getAsJsonObject("experimental");
            experimental.addProperty(property, value);
            break;
          }
        }
```

you will be able to do:
```kotin
        String property = URLDecoder.decode(query.substring(0, idx), UTF_8);
        String value = URLDecoder.decode(query.substring(idx + 1), UTF_8);

        boolean isExperimental = false;
        if (experimentalParameters.contains(property)) {
          isExperimental = true;
          if (!optionsJson.has("experimental")) {
            optionsJson.add("experimental", new JsonObject());
          }
          JsonObject experimental = optionsJson.getAsJsonObject("experimental");
          experimental.addProperty(property, value);
        }
        
```

Also runtime complexity of lookup in hash table is O(1) vs O(n) in array. But I don't think performance is a good reason for this change, because array has small and constant size.

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/ExperimentalWaypointMetadata.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  public abstract String plugType();
+
+  /**
+   * Object representing experimental value.
+   * <p>
+   * All available experimental values are subject to change at any time.
+   */
+  @Nullable
+  @SerializedName("power_kw")
+  public abstract Integer powerKiloWatt();
```
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/ExperimentalWaypointMetadata.java:82]: Thanks for bringing this up :+1: 

This response data will only be returned if the `electric engine` experimental parameter is used. Since it shouldn't be used in any type of production environment, we shouldn't need to worry about compatibility for our case. 

That said, I don't want to setup a dangerous precedent where an "experimental" property would be introduced and released in a stable version of the library and then potentially conflict with a server side response much, much later down the line, causing parsing exceptions.

I will take another pass at this, to make this safer. My current idea is to introduce a new annotation we discussed above but not only for documentation reasons but also to ensure that a stable version of `mapbox-java` cannot be released with an experimental property. Planning to use the annotation processing and `BuildConfig` to see if I can throw a build-time (preferably) or runtime exception in these cases. If this won't work, then maybe some alternative standalone script.

**Review by VysotskiVadim — APPROVED** [review]
I'm okay if this changes will be released in not stable version of `mapbox-java`. The comments I left could be a blocker (maybe not 😄 ) only for a stable release.

**Comment by LukasPaczos** [comment]
I'm going ahead and merging here. I've been trying to figure out some automated way of verifying the existence of experimental features, or gracefully falling back when they fail to parse, but couldn't find any in the timeboxed exercise.

Similarly to https://github.com/mapbox/mapbox-java/issues/1344, we'd need to change how the Auto-Value-Gson lib works to easily fallback to `null` if a property fails to parse. If we wanted custom type adapters without modifying the generated code, we'd most-likely also have to have custom return types (even for primitives) which isn't ideal either.

Anyway, the plan would be to revert this PR before the next stable release and budget more time for looking for other options to maintain experimental properties.
